### PR TITLE
Modifier style constructor calls without parentheses are an error.

### LINKED ISSUE #3325: SOL-006 Duplicated super-constructor calls not reported
**Description**

Solidity has two methods to define super constructor calls: the first one is embedding the arguments in super contract definition and the second is by declaring a new constructor and adding the name and arguments for the super contract. If both methods are used concurrently, the constructor definition is used. However, the compiler does not warn the user that the first set of arguments has been overridden by the second.
The following example shows the problem. The parent constract is constructed with the
argument 40, not 20.

```
pragma solidity ^0.4.16;

contract P1 {
constructor P1(uint v) {}
}
contract P2 is​ ​P1(20)​ {
function P2(uint v)​ ​P1(40)​ {
}
}
```

**Recommendations**

Either prevent the user from defining two super-constructors or warn if one super-constructor overrides the other.

### LINKED ISSUE #3838: Modifier-style constructor calls without parentheses should be an error.
Came up in #3821.

In this code snipped:

    contract A { constructor() public { } }
    contract B1 is A { constructor() A() public { } }
    contract B2 is A { constructor() A public { } }

the constructor call in ``B1`` is fine, but in ``B2`` the ``À`` in ``constructor() A public {}`` serves no purpose and should be disallowed.

### PR DESCRIPTION
**Pull request by ekpyron**
Closes #3325.
Closes #3838.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ Features:
  * Syntax Checker: Issue warning for empty structs (or error as experimental 0.5.0 feature).
  * General: Introduce new constructor syntax using the ``constructor`` keyword as experimental 0.5.0 feature.
  * Inheritance: Error when using empty parenthesis for base class constructors that require arguments as experimental 0.5.0 feature.
+ * Inheritance: Error when using no parenthesis in modifier-style constructor calls as experimental 0.5.0 feature.
```
Comment (chriseth) [inline Changelog.md]: `parentheses`?

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 2]
```diff
@@ -378,7 +378,7 @@ bool ASTJsonConverter::visit(ModifierInvocation const& _node)
 {
 	setJsonNode(_node, "ModifierInvocation", {
 		make_pair("modifierName", toJson(*_node.name())),
-		make_pair("arguments", toJson(_node.arguments()))
+		make_pair("arguments", _node.arguments() ? toJson(*_node.arguments()) : Json::Value(Json::arrayValue))
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp]: Perhaps it should be `null` if the arguments are not given.

**Review by chriseth — DISMISSED** [review]

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 3]
```diff
@@ -268,7 +268,7 @@ bool ASTJsonConverter::visit(InheritanceSpecifier const& _node)
 {
 	setJsonNode(_node, "InheritanceSpecifier", {
 		make_pair("baseName", toJson(_node.name())),
-		make_pair("arguments", _node.arguments() ? toJson(*_node.arguments()) : Json::Value(Json::arrayValue))
+		make_pair("arguments", _node.arguments() ? toJson(*_node.arguments()) : Json::nullValue)
```
Comment (ekpyron) [inline libsolidity/ast/ASTJsonConverter.cpp:271]: I also changed to ``Json::nullValue`` in the absence of arguments here, although this should actually have been done in #3824.

Comment (chriseth) [reply libsolidity/ast/ASTJsonConverter.cpp:271]: Great, thanks!

**Review by chriseth — APPROVED** [review]

**Pull request merged**